age: add page (#5758)
tldr-pages/tldr: pages/common/age.md

# age

> A simple, modern and secure file encryption tool.
> More information: <https://age-encryption.org>.

- Generate an encrypted file that can be decrypted with a passphrase:

> `age -p -o {{/path/to/encrypted_file}} {{/path/to/unencrypted_file}}`

Reviewer: ```suggestion
`age --passphrase --output {{/path/to/encrypted_file}} {{/path/to/unencrypted_file}}`
```

- Generate a keypair, saving the private key and the public key to an (unencrypted) keyfile, and printing the public key to stdout:

> `age-keygen -o {{/path/to/keyfile}}`

Reviewer: ```suggestion
`age-keygen --output {{/path/to/keyfile}}`
```

- Encrypt a file to one or several public keys that are entered as literals:

> `age -o {{/path/to/encrypted_file}} -r {{publickey1}} -r {{publickey2}} {{/path/to/unencrypted_file}}`

Reviewer: ```suggestion
`age --output {{/path/to/encrypted_file}} --recipient {{publickey1}} --recipient {{publickey2}} {{/path/to/unencrypted_file}}`
```

> - Encrypt a file to one or several public keys that are specified in a recipients file, with each public key separated by a newline, lines starting with `#` being ignored:

Reviewer: ```suggestion
- Encrypt a file to one or several public keys that are specified in a recipients file:
```

> `age -o {{/path/to/encrypted_file}} -R {{/path/to/recipients_file}} {{/path/to/unencrypted_file}}`

Reviewer: ```suggestion
`age --output {{/path/to/encrypted_file}} --recipients-file {{/path/to/recipients_file}} {{/path/to/unencrypted_file}}`
```

# age

> A simple, modern and secure file encryption tool.
> More information: <https://age-encryption.org>.

- Generate an encrypted file that can be decrypted with a passphrase:

`age --passphrase --output {{path/to/encrypted_file}} {{path/to/unencrypted_file}}`

> - Generate a keypair, saving the private key and the public key to an unencrypted keyfile and printing the public key to stdout:

Reviewer 2: ```suggestion
- Generate a key pair, saving the private key to an unencrypted file and printing the public key to stdout:
```

keypair and keyfile are not really words, and while the public key does end up in a comment in the output, I think it's a little confusing to say they are both saved.

Reviewer 3: Hey there. Thanks for the review. In other pages we use both `keypair` and `key pair`, so I would suggest here that we go with whichever way the official documentation for the command does it.

Author: Well, the official documentation doesn't specify anything AFAICT, but @FiloSottile is the main developer of Age, so I guess we can trust him in this regard?

Reviewer 4: Yes, let's do that. It's also better to use real words whereever possible 👍🏻

Reviewer 3: Oh, cool! I didn't realise. Sounds good then, let's do this.

> - Generate a keypair, saving the private key and the public key to an (unencrypted) keyfile, and printing the public key to stdout:

Reviewer 4: ```suggestion
- Generate a keypair, saving the private key and the public key to an unencrypted keyfile and printing the public key to stdout:
```

`age-keygen --output {{path/to/keyfile}}`

- Encrypt a file to one or several public keys that are entered as literals:

> `age --output {{path/to/encrypted_file}} --recipient {{publickey1}} --recipient {{publickey2}} {{path/to/unencrypted_file}}`

Reviewer 2: Here and in the `--recipients-file` example below, it would be nice for consistency to move `--output` to be the last argument.

> - Encrypt a file to one or several public keys that are entered as literals:
> 
> `age --output {{path/to/encrypted_file}} --recipient {{publickey1}} --recipient {{publickey2}} {{path/to/unencrypted_file}}`

Reviewer 4: ```suggestion
- Encrypt a file with one or more public keys:

`age --output {{path/to/encrypted_file}} --recipient {{public_key1}} --recipient {{public_key2}} {{path/to/unencrypted_file}}`
```

Author: I'm not too sure about those… Is encrypting "with" a public key indeed better than "to"? Perhaps it might be easier to understand by new users, perhaps not… And I'd rather like to keep the "as literals" part in order to make clear that these aren't supposed to be files… YMMV, of course.

Author: I implemented some of your suggestions now, just not the ones I expressed concerns with above. Let me know what you think.

Reviewer 4: Hmm, that sounds ok to me.  "Encrypt a file to a key" sounds odd to me, but that might be because I'm not that familiar with encryption myself, having done little of it.  If you're confident that it's technically correct, I'm perfectly fine with about it. 

About "as literals" - I know what you mean; it is a possible concern, but we normally add ornaments such as `path/to/`, file extensions, or something like `public_key_file` to denote that it is a file and not a string.  I don't mind the current wording myself, but I searched our 2,339 English pages for "literal" and only found 6 pages...but that doesn't mean we can't do it now.

@sbrl, could we get some advice on this?

Reviewer 3: Hey there! I agree with @bl-ue here that `Encrypting to a key` really doesn't make a lot of sense here. `with` indicates that it uses  the public key in the process, whereas `to` would imply that it converts the input _into_ a public key, which does not make sense.

Where does it get the public keys from? If it's the id fo a key it pulls say from GPG, then `public_key_id_1` and `public_key_id_2` would be more appropriate. If the public key is stored in a file on disk that is referenced, then prefixing it with `path/to/` would be preferable to ensure consistency with other tldr pages.

Author: It's the literal Base64 encoded public key. Ed25519 ones are short enough to do this. Age is designed as a partial replacement for PGP/GPG, and has nothing to do with those. For specifying a file with the public keys, look at the example below.

Author: > Public key encryption, in which a message is encrypted with a recipient's public key.

https://en.wikipedia.org/wiki/Public-key_cryptography#Description

Wikipedia seems to agree with you two, btw. I'll just change it to that, then.

Author: @sbrl What I could do is specify that it should be Base64 encoded (as was done in Minisign), however, that's the format one would always get Age keys in, thus making that perhaps a bit redundant, and it's also not really mentioned in the official instructions (both in the GitHub README and in `age --help`)… Perhaps @FiloSottile would like to chime in here?

Reviewer 3: Cool, thanks.

Hrm. Given that this command both generates and uses keypairs, I would suggest that the detail that the keys are stored as base64 isn't something that needs to be mentioned in the tldr page. Remember that the page is supposed to serve as a quick reference guide - it's not meant to be exhaustive :-)

Reviewer 2: Users are not expected to think about the encoding of the recipients (and it's technically not Base64), so I would avoid mentioning it.

- Encrypt a file to one or several public keys that are specified in a recipients file:

`age --output {{path/to/encrypted_file}} --recipients-file {{path/to/recipients_file}} {{path/to/unencrypted_file}}`

- Decrypt a file with a passphrase:

`age --decrypt --output {{path/to/decrypted_file}} {{path/to/encrypted_file}}`

> - Decrypt a file with a private keyfile:

Reviewer 2: ```suggestion
- Decrypt a file with a private key:
```

Reviewer 3: If it is loading the private key from a file, then it should be either `key file` or `keyfile`, not just simply `key`.

Author: The official documentation refers to it as "identity file", but that might be confusing for users not familiar with Age, at least it was a somewhat weird term for me. `key file` should work, I guess.
